python/lsst/pipe/base/_check_qg_outputs.py

```diff
@@ -0,0 +1,216 @@
+from lsst.resources import ResourcePathExpression
+
+from .graph import QuantumGraph
+from .pipeline import PipelineDatasetTypes
+
+
+@dataclasses.dataclass
+class DatasetTypeExecutionReport:
+    missing_failed: set[DatasetRef] = dataclasses.field(default_factory=set)
+    """Datasets not produced because their quanta failed directly in this run.
+    """
+    missing_not_produced: dict[DatasetRef, bool] = dataclasses.field(default_factory=dict)
+    """Datasets not produced because their inputs were not produced or not
+    found
+    """
+    # bool: predicted inputs to this task were not produced
+    missing_upstream_failed: set[DatasetRef] = dataclasses.field(default_factory=set)
+    """Datasets not produced due to an upstream failure
+    """
+    n_produced: int = 0
+    """Counts of datasets produced by this run.
+    """
+
+    def to_summary_dict(self) -> dict[str, Any]:
+        return {
+            "produced": self.n_produced,
+            "missing_failed": len(self.missing_failed),
+            "missing_not_produced": len(self.missing_not_produced),
+            "missing_upstream_failed": len(self.missing_upstream_failed),
+        }
+
+    def handle_missing_dataset(self, output_ref: DatasetRef, failed: bool, status_graph: networkx.DiGraph):
+        if failed:
+            for upstream_quantum_id in status_graph.predecessors(output_ref.id):
+                if status_graph.nodes[upstream_quantum_id]["failed"]:
+                    self.missing_upstream_failed.add(output_ref)
+                    break
+            else:
+                self.missing_failed.add(output_ref)
+        else:
+            status_graph.nodes[output_ref.id]["not_produced"] = True
+            self.missing_not_produced[output_ref] = any(
+                status_graph.nodes[upstream_dataset_id].get("not_produced", False)
+                for upstream_quantum_id in status_graph.predecessors(output_ref.id)
+                for upstream_dataset_id in status_graph.predecessors(upstream_quantum_id)
+            )
+
+    def handle_produced_dataset(self, output_ref: DatasetRef, status_graph: networkx.DiGraph):
+        status_graph.nodes[output_ref.id]["not_produced"] = False
+        self.n_produced += 1
+
+
+@dataclasses.dataclass
+class TaskExecutionReport:
+    failed: dict[uuid.UUID, DatasetRef] = dataclasses.field(default_factory=dict)
+    """A mapping from quantum data ID to log dataset reference for quanta that
+    failed directly in this run.
+    """
+    failed_upstream: dict[uuid.UUID, DataCoordinate] = dataclasses.field(default_factory=dict)
+    """A mapping of data IDs of quanta that were not attempted due to an
+    upstream failure.
+    """
+    output_datasets: dict[str, DatasetTypeExecutionReport] = dataclasses.field(default_factory=dict)
+    """Reports of the missing and produced outputs of each DatasetType
+    """
+
+    def inspect_quantum(self, quantum_node, status_graph: networkx.DiGraph, refs, metadata_name, log_name):
```

Comment:
Needs a docstring, and also some type annotations:
```suggestion
    def inspect_quantum(
        self,
        quantum_node: QuantumNode,
        status_graph: networkx.DiGraph,
        refs: Iterable[DatasetRef],
        metadata_name: str,
        log_name: str,
    ) -> None:
```
This will require
```
from .graph import QuantumNode
```
at the top.